### PR TITLE
Make Two Factor Field numeric

### PR DESCRIPTION
**Pull request by nupplaphil**
On mobile phones, you automatically will now have a numeric-only keyboard for the TOTP input :-)

### CODE REVIEW

File: view/templates/field_input.tpl [hunk 1]
```diff
@@ -1,7 +1,7 @@
 	
 	<div class='field input' id='wrapper_{{$field.0}}'>
 		<label for='id_{{$field.0}}'>{{$field.1}}</label>
-		<input{{if $field.6 eq 'email'}} type='email'{{elseif $field.6 eq 'url'}} type='url'{{else}} type="text"{{/if}} name='{{$field.0}}' id='id_{{$field.0}}' value="{{$field.2 nofilter}}"{{if $field.4 eq 'required'}} required{{/if}}{{if $field.5 eq "autofocus"}} autofocus{{elseif $field.5}} {{$field.5 nofilter}}{{/if}} aria-describedby='{{$field.0}}_tip'>
+		<input{{if $field.6 eq 'email'}} type='email'{{elseif $field.6 eq 'url'}} type='url'{{elseif $field.6 eq 'int'}} type='tel'{{else}} type="text"{{/if}} name='{{$field.0}}' id='id_{{$field.0}}' value="{{$field.2 nofilter}}"{{if $field.4 eq 'required'}} required{{/if}}{{if $field.5 eq "autofocus"}} autofocus{{elseif $field.5}} {{$field.5 nofilter}}{{/if}} aria-describedby='{{$field.0}}_tip'>
```
Comment (tobiasd) [inline view/templates/field_input.tpl]: Could you please update the documentation about the templates with this addition ;-)

Comment (nupplaphil) [reply view/templates/field_input.tpl]: aye aye :-) tbh I didn't know there's a doc about it ^^

Comment (MrPetovan) [inline view/templates/field_input.tpl]: Please convert all single quotes for attribute values to double-quotes.

Comment (MrPetovan) [inline view/templates/field_input.tpl]: Shouldn't we use "number" instead?

Comment (nupplaphil) [reply view/templates/field_input.tpl]: mhm .. you're right .. I searched for the main difference between tel and numeric and numeric is better for this type of input, yes

Comment (MrPetovan) [reply view/templates/field_input.tpl]: My worry is for leading 0s to still be accepted as input for a number, other than that it sounds more appropriate.

File: view/templates/field_input.tpl [hunk 2]
```diff
@@ -1,8 +1,8 @@
 	
-	<div class='field input' id='wrapper_{{$field.0}}'>
-		<label for='id_{{$field.0}}'>{{$field.1}}</label>
-		<input{{if $field.6 eq 'email'}} type='email'{{elseif $field.6 eq 'url'}} type='url'{{elseif $field.6 eq 'int'}} type='tel'{{else}} type="text"{{/if}} name='{{$field.0}}' id='id_{{$field.0}}' value="{{$field.2 nofilter}}"{{if $field.4 eq 'required'}} required{{/if}}{{if $field.5 eq "autofocus"}} autofocus{{elseif $field.5}} {{$field.5 nofilter}}{{/if}} aria-describedby='{{$field.0}}_tip'>
+	<div class="field input" id="wrapper_{{$field.0}}">
+		<label for="id_{{$field.0}}">{{$field.1}}</label>
+		<input{{if $field.6 eq 'email'}} type="email"{{elseif $field.6 eq 'url'}} type="url"{{elseif $field.6 eq 'int'}} type="number"{{else}} type="text"{{/if}} name="{{$field.0}}" id="id_{{$field.0}}" value="{{$field.2 nofilter}}"{{if $field.4 eq 'required'}} required{{/if}}{{if $field.5 eq "autofocus"}} autofocus{{elseif $field.5}} {{$field.5 nofilter}}{{/if}} aria-describedby="{{$field.0}}_tip">
```
Comment (MrPetovan) [inline view/templates/field_input.tpl]: Please use the actual value of `$field.6` for `type` or `text` if it isn't set. This means changing `int` to `number` in the possible values.

File: doc/smarty3-templates.md [hunk 3]
```diff
@@ -89,7 +89,7 @@ Field parameter:
 3. Help text for the input box,
 4. if set to "required" modern browser will check that this input box is filled when submitting the form,
 5. if set to "autofocus" modern browser will put the cursur into this box once the page is loaded,
-6. if set to "email", "url" or "int" modern browser will check that the filled in value corresponds to an email address, URL or numeric.
+6. if set, it will be used for the input type, default is `text` (possible types: https://www.w3schools.com/html/html_form_input_types.asp).
```
Comment (MrPetovan) [inline doc/smarty3-templates.md]: NO NO NO NO W3SCHOOLS LINKS EVER ON MY WATCH!

They are leechers using sometimes out-of-date info to push ads and SEO their way up on Google. Use the MDN instead: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#%3Cinput%3E_types

File: view/templates/field_input.tpl [hunk 4]
```diff
@@ -1,7 +1,7 @@
 	
 	<div class="field input" id="wrapper_{{$field.0}}">
 		<label for="id_{{$field.0}}">{{$field.1}}</label>
-		<input{{if $field.6 eq 'email'}} type="email"{{elseif $field.6 eq 'url'}} type="url"{{elseif $field.6 eq 'int'}} type="number"{{else}} type="text"{{/if}} name="{{$field.0}}" id="id_{{$field.0}}" value="{{$field.2 nofilter}}"{{if $field.4 eq 'required'}} required{{/if}}{{if $field.5 eq "autofocus"}} autofocus{{elseif $field.5}} {{$field.5 nofilter}}{{/if}} aria-describedby="{{$field.0}}_tip">
+		<input{{if $field.6}} type="{{$field.6}}"{{else}} type="text"{{/if}} name="{{$field.0}}" id="id_{{$field.0}}" value="{{$field.2 nofilter}}"{{if $field.4 eq 'required'}} required{{/if}}{{if $field.5 eq "autofocus"}} autofocus{{elseif $field.5}} {{$field.5 nofilter}}{{/if}} aria-describedby="{{$field.0}}_tip">
```
Comment (MrPetovan) [inline view/templates/field_input.tpl]: Suggestion: `type="{{$field.6 ?? 'text'}}"`

Comment (nupplaphil) [reply view/templates/field_input.tpl]: that works? wow .. I'm more and more learning ^^ .. I now know why I normally better stay at the backend ^^

File: doc/smarty3-templates.md (hunk 3 above)
Comment (nupplaphil) [reply doc/smarty3-templates.md]: .. oh .. good 2 know ..

File: view/templates/field_input.tpl (hunk 4 above)
Comment (MrPetovan) [reply view/templates/field_input.tpl]: Please test first, Smarty template allow a bunch of PHP syntax, which makes me believe we would be better off using pure PHP templates.

Comment (nupplaphil) [reply view/templates/field_input.tpl]: my local instance cries:
```console
replaceMacros: Syntax error in template "file:/vagrant/view/templates/field_input.tpl"  on line 4 ""  - Unexpected " ?", expected one of: "}}"
```

Comment (nupplaphil) [reply view/templates/field_input.tpl]: I know from Level-2 (who invented Dice) that they introduced a "sample like" template engine. So you create a plain HTML doc with sample-text and than the engine replaces the sample-text during rendering. It's mainly build for frontend-engineers so they can create new pages without the need to know about PHP-internal stuff ;-) .. https://github.com/Level-2/Transphporm

Comment (MrPetovan) [reply view/templates/field_input.tpl]: Maybe `$field.6 ? $field.6 : 'text'` ?

Comment (MrPetovan) [reply view/templates/field_input.tpl]: > It's mainly build for frontend-engineers so they can create new pages without the need to know about PHP-internal stuff

I still don't believe these exist. It's been the justification for every single templating engine and I've never seen anyone use a templating engine without also doing the underlying PHP backend which is super-dumb.

Comment (nupplaphil) [reply view/templates/field_input.tpl]: > > It's mainly build for frontend-engineers so they can create new pages without the need to know about PHP-internal stuff
> 
> I still don't believe these exist. It's been the justification for every single templating engine and I've never seen anyone use a templating engine without also doing the underlying PHP backend which is super-dumb.

just have a look at the link :) At least they drop the whole variable-stuff out of the plain templates and overwrite the values at the CSS base.

Comment (nupplaphil) [reply view/templates/field_input.tpl]: > Maybe `$field.6 ? $field.6 : 'text'` ?

computer says no :-) I let it as is to not brick something..

File: view/templates/field_input.tpl (hunk 1 above)
Comment (tobiasd) [reply view/templates/field_input.tpl]: I know, devs and the "D" word. You should meet @annando next hackathon in Berlin so you two can argue against the "D" word issues :-D

File: view/templates/field_input.tpl (hunk 4 above)
Comment (MrPetovan) [reply view/templates/field_input.tpl]: > just have a look at the link :) At least they drop the whole variable-stuff out of the plain templates and overwrite the values at the CSS base.

I did, and now I want to gouge my eyes out. It basically moves code logic from templates into having presentation logic in the code, at the price of having an additional intermediate template file, which isn't ideal either.

Comment (nupplaphil) [reply view/templates/field_input.tpl]: At least it fully seperate the design from the template engine itself ;) but it just came across my way during some search

**Comment by nupplaphil** [comment]
@mrpetovan I think we should switch to "tel" because on mobile, tel fields triggers the tel keyboard which contains additional buttons like the copy&pasted button. The native number keyboard really just shows the 0-9 keyboards without any other button.

File: view/templates/field_input.tpl (hunk 4 above)
Comment (MrPetovan) [reply view/templates/field_input.tpl]: This is true, but you need the design mock-up to be actually functional for this to be true. And if the design mock-up changes, the TSS files have to be updated, with a high potential for dead selectors as with CSS... Ugh Ugh Ugh.

**Pull request merged**